Support GPU Simulations

GeorgeR227 opened the pull request:
Adding CUDA support into Decapodes. This adds a code target keyword into gensim that allows a user to output code that can target either the CPU or a Nvidia GPU with CUDA. This is done in a CUDA extension file.

GeorgeR227 commented:
Assuming that https://github.com/AlgebraicJulia/CombinatorialSpaces.jl/pull/75 isn't altered in any drastic way, and it shouldn't be at this point, this PR is ready for review.

lukem12345 commented on docs/src/cahn_hilliard.md:
````diff
@@ -0,0 +1,99 @@
+to_graphviz(CahnHilliard)
+```
+
+In these equations, C will represent the concentration of the binary fluid, range from -1 to 1 to differentiate between different phase. Then we have D as a diffusion constant and γ as the length of the transition regions. This formulation of the Cahn-Hilliard equations was drawn from the Wikipedia page on the topic, found here: https://en.wikipedia.org/wiki/Cahn%E2%80%93Hilliard_equation.
````
Isn't the square root of this constant the transition length?

GeorgeR227 replied:
Yes, I've cleaned up the wording here.

lukem12345 commented on examples/diff_adv/source_heat.jl:
```diff
@@ -0,0 +1,122 @@
+sd = EmbeddedDeltaDualComplex2D{Bool, Float64, Point3D}(s);
+subdivide_duals!(sd, Circumcenter());
+
+function generate(sd, symbol, hodge=DiagonalHodge())
```
Delete and use nothing

lukem12345 commented on examples/diff_adv/source_heat.jl:
```diff
@@ -0,0 +1,122 @@
+using Catlab
```
Why catlab

GeorgeR227 replied:
Needed for ```nparts```, I'm going to delete this file anyway since this was just experimental and doesn't actually work.

lukem12345 commented on examples/diff_adv/source_heat.jl:
```diff
@@ -0,0 +1,122 @@
+constants_and_parameters = (
+  c = 5000 * 1000,
+  Dif = 20,
+  S = t -> S * t
```
Confusing, rename the cuvector

lukem12345 commented on src/Decapodes.jl:
```diff
@@ -17,8 +17,9 @@ using DiagrammaticEquations.Deca
-CartesianPoint, SpherePoint, r, theta, phi, TangentBasis, θhat, ϕhat  
+gensim, evalsim, closest_point, findnode, compile, compile_env,  PhysicsState,  default_dec_matrix_generate, default_dec_cu_matrix_generate, default_dec_generate, VectorForm,
+CartesianPoint, SpherePoint, r, theta, phi, TangentBasis, θhat, ϕhat,
+cpu, cuda
```
Pick better names for these types

GeorgeR227 replied:
I think these types are nice and they match with similar types used for device selection in something like PyTorch. Did you have anything else in mind?

jpfairbanks replied:
Types should at least have capital names like CPU and CUDA

GeorgeR227 replied:
Originally those were the names I was using but CUDA name conflicts with the package name, also CUDA, so that won't work well. I could change it to be ```Cpu``` and ```Cuda```.

lukem12345 replied:
Can have an abstract type for CompilationTarget and subtype CUDATarget CPUTarget

GeorgeR227 replied:
I already do that here: https://github.com/AlgebraicJulia/Decapodes.jl/blob/6ae30c6efecea9b7c6b536bc9719368a735255ab/src/simulation.jl#L14-L17

lukem12345 commented on src/simulation.jl:
```diff
@@ -101,19 +109,28 @@ Base.Expr(c::AllocVecCall) = begin
-elseif(type == :Form2)
-return 2
+# TODO: Allow user to overload these hooks with user-defined code_target
+function hook_ExprAVC_generate_cache_expr(c::AllocVecCall, resolved_form::Symbol, ::cuda)
```
This name is too long

GeorgeR227 replied:
Sure but this is not exposed to the user, is still descriptive and can be changed when the ```TODO``` is done.

GeorgeR227 replied:
I've changed the hook names to be shorter, in the format ```hook_$(FunctionAcronymWhereCalled)_$(BehaviorBeingModified)```

lukem12345 commented on src/simulation.jl:
```diff
@@ -164,18 +181,24 @@ function compile_env(d::AbstractNamedDecapode, dec_matrices::Vector{Symbol}, con
     # Could turn this into a special call
-    def = :(($mat_op, $op) = default_dec_matrix_generate(mesh, $quote_op, hodge))
+
+    default_generation = :default_dec_matrix_generate
```
Use ternary operator

GeorgeR227 replied:
Can't since we could be using either in-place or out-of-place in the same simulation. Also this would break the way default_dec_matrix_generate works.

GeorgeR227 replied:
Oh, I didn't realize you meant choosing the default generation. I've updated this to an MLStyle in-case we want to add more targets. Also I've TODOed this line so when we allow users to overload hooks with their own code targets, we can update this.

lukem12345 commented on src/simulation.jl:
```diff
@@ -483,9 +514,28 @@ function link_contract_operators(d::SummationDecapode, con_dec_operators::Set{Sy
+  return :($(add_inplace_stub(computation_name)) = $(Expr(:call, :*, computation...)))
+end
+
+function generate_parentheses_multipy(list)
```
Comment to explain what is happening here

GeorgeR227 replied:
Done.

lukem12345 commented on src/simulation.jl:
```diff
@@ -494,7 +544,7 @@ function gensim(user_d::AbstractNamedDecapode, input_vars; dimension::Int=2,
   #d′ = average_rewrite(d′)
 
-  replace_negation_with_multiply!(d′, input_vars)
+  # replace_negation_with_multiply!(d′, input_vars)
```
Delete?

GeorgeR227 replied:
Done.

lukem12345 commented on src/simulation.jl:
```diff
@@ -513,23 +563,22 @@ function gensim(user_d::AbstractNamedDecapode, input_vars; dimension::Int=2,
   # This will generate all of the fundemental DEC operators present
-  optimizable_dec_operators = Set([:⋆₀, :⋆₁, :⋆₂, :⋆₀⁻¹, :⋆₂⁻¹,
-    :d₀, :d₁, :dual_d₀, :d̃₀, :dual_d₁, :d̃₁])
+  optimizable_dec_operators = Set([:⋆₀, :⋆₁, :⋆₂, :⋆₀⁻¹, :⋆₂⁻¹, :d₀, :d₁, :dual_d₀, :d̃₀, :dual_d₁, :d̃₁])
```
Reason for this change?

GeorgeR227 replied:
No, reverted to similar form.

lukem12345 commented on src/simulation.jl:
```diff
@@ -553,11 +602,10 @@ gensim(collate(c); dimension=dimension)
 """
-gensim(d::AbstractNamedDecapode; dimension::Int=2, stateeltype = Float64) = gensim(d,
-                                                                                   vcat(collect(infer_state_names(d)), d[incident(d, :Literal, :type), :name]), dimension=dimension, stateeltype=stateeltype)
+gensim(d::AbstractNamedDecapode; dimension::Int=2, stateeltype = Float64, code_target = cpu()) = gensim(d, vcat(collect(infer_state_names(d)), d[incident(d, :Literal, :type), :name]), dimension=dimension, stateeltype=stateeltype, code_target=code_target)
```
Line too long

lukem12345 commented on src/simulation.jl:
```diff
@@ -553,11 +602,10 @@ gensim(collate(c); dimension=dimension)
 
-evalsim(d::AbstractNamedDecapode; dimension::Int=2, stateeltype = Float64) = eval(gensim(d, dimension=dimension, stateeltype=stateeltype))
-evalsim(d::AbstractNamedDecapode, input_vars; dimension::Int=2, stateeltype = Float64) = eval(gensim(d, input_vars, dimension=dimension, stateeltype=stateeltype))
+evalsim(d::AbstractNamedDecapode; dimension::Int=2, stateeltype = Float64, code_target = cpu()) = eval(gensim(d, dimension=dimension, stateeltype=stateeltype, code_target=code_target))
```
Too long

GeorgeR227 replied:
Fixed for now, we should just pass all keywords in the coming compiler update.

lukem12345 commented on test/componentarrays.jl:
```diff
@@ -93,6 +93,40 @@ soln(0.9)
 #     ob.color = findnode(soln(t), :C)[point_map]
 # end
 
+function closest_point(p1, p2, dims)
```
Delete

GeorgeR227 replied:
Can't this is used in flat_op

GeorgeR227 replied:
Removing these should become an issue and a separate PR. I would consider removing this not in the scope of this PR.

lukem12345 commented on test/componentarrays.jl:
```diff
@@ -93,6 +93,40 @@ soln(0.9)
+  Point3{Float64}(p_res...)
+end
+
+function flat_op(s::AbstractDeltaDualComplex2D, X::AbstractVector; dims=[Inf, Inf, Inf])
```
Delete

GeorgeR227 replied:
Can't this is used in line: https://github.com/AlgebraicJulia/Decapodes.jl/blob/aea31c661808c5d1aed571af70c1bc88bb3d0bf7/test/componentarrays.jl#L152

GeorgeR227 replied:
Removing these should become an issue and a separate PR. I would consider removing this not in the scope of this PR.

lukem12345 commented on test/cuda_sims.jl:
```diff
@@ -0,0 +1,344 @@
+@testset "Heat Equation Float32" begin
+  Heat = @decapode begin
+    U::Form0
+    ∂ₜ(U) == 100 * Δ(U)
```
100f0

GeorgeR227 replied:
True but the gensim will convert this to a Float32 anyway, I've checked the output and can confirm.

lukem12345 commented on test/cuda_sims.jl:
```diff
@@ -0,0 +1,344 @@
+end
+
+@testset "Brusselator Float64" begin
+  Brusselator = @decapode begin
```
Canon

GeorgeR227 replied:
Turn this into an issue and another PR since there are tests in other files that should also use canon.

lukem12345 commented on test/cuda_sims.jl:
```diff
@@ -0,0 +1,344 @@
+  cuda_lap_mat = CuSparseMatrixCSC(lap_mat)
+  function cuda_generate(sd, my_symbol; hodge=DiagonalHodge())
+    op = @match my_symbol begin
+      :Δ => x -> begin
```
Is this included now

GeorgeR227 replied:
No, this is a dual-dual Laplace so this is not in ```default``` set. This should however be an issue, along with other operators.

GeorgeR227 commented on examples/diff_adv/wave.jl:
```diff
@@ -0,0 +1,73 @@
+    ∂ₜ(U) == T
+end
+
+function simulate(mesh, operators, hodge = GeometricHodge())
```
Can't since I don't have support for second time derivatives right now but I can remove it from this PR and have it ready for when that does work.

GeorgeR227 commented on ext/DecapodesCUDAExt.jl:
```diff
@@ -0,0 +1,125 @@
+end
+
+# TODO: Update this to better cast hodges
+function dec_cu_mat_hodge(k, sd::HasDeltaSet, hodge)
```
I think the issue here was that I should be multi-dispatching on ```dec_mat_hodge```, between CPU and CUDA version, but I think this is moot since this version seems preferable to me.

GeorgeR227 commented on ext/DecapodesCUDAExt.jl:
```diff
@@ -0,0 +1,125 @@
+
+# Special case for inverse hodge for DualForm1 to Form1
+# TODO: This should be changed to use Krylov, need to figure out inplace version of this
+function dec_cu_pair_inv_hodge(::Type{Val{1}}, sd::EmbeddedDeltaDualComplex2D{Bool, float_type, _p} where _p,  ::GeometricHodge ) where float_type
```
Sure but this function will not be exposed to the user.

GeorgeR227 replied:
Done.

GeorgeR227 commented on ext/DecapodesCUDAExt.jl:
```diff
@@ -0,0 +1,125 @@
+end
+
+# Special case for inverse hodge for DualForm1 to Form1
+# TODO: This should be changed to use Krylov, need to figure out inplace version of this
```
Yeah this is done and works.

GeorgeR227 commented on examples/chemistry/gompertz.jl:
Done.

GeorgeR227 commented on examples/climate/klausmeier.jl:
Removed from examples.

GeorgeR227 commented on examples/diff_adv/cht.jl:
Done.

GeorgeR227 commented on examples/diff_adv/heat.jl:
```diff
@@ -0,0 +1,64 @@
+using MLStyle
+using ComponentArrays
+using OrdinaryDiffEq
+# using CairoMakie
```
The heat.jl file has been updated.

GeorgeR227 commented on examples/diff_adv/wave.jl:
```diff
@@ -0,0 +1,90 @@
+    subdivide_duals!(sd, Circumcenter())
+end
+
+Wave = @decapode begin
```
This example will be moved onto another branch.

GeorgeR227 commented:
@lukem12345 I've fixed the most relevant issues you brought up. I don't want merging this branch to get delayed behind examples, when there are other examples anyway that are old/broken, or documentation, that doesn't work with CUDA anyway, so I've removed any with issues and made a new branch with them we can fix and merge later.

lukem12345 commented on examples/diff_adv/burger.jl:
```diff
@@ -18,6 +18,8 @@ using MLStyle
 using ComponentArrays
 using OrdinaryDiffEq
 
+## Original implementation can be found at: https://github.com/AlgebraicJulia/DECAPODES-Benchmarks
```
Comment on wrong file

GeorgeR227 replied:
That's true, now only CHT and CHT_cuda have this link.

GeorgeR227 commented on docs/src/cahn_hilliard.md:
That's true, that's been fixed.

GeorgeR227 commented on docs/src/cahn_hilliard.md:
```diff
@@ -0,0 +1,99 @@
+fig = Figure()
+ax = CairoMakie.Axis(fig[1,1], aspect=1)
+wf = wireframe!(ax, s)
+save("CahnHillard_Rect.png", fig)
```
Fixed. I think I had that behavior implemented at one point but never pushed.

GeorgeR227 commented on docs/src/cahn_hilliard.md:
````diff
@@ -0,0 +1,99 @@
+constants = (D = 0.5, γ = 0.5);
+```
+
+We'll now create the simulation code representing the Cahn-Hilliard equations. We pass nothing in the second arguement to sim since we have no custom functions to pass in.
````
Good argument, fixed.

GeorgeR227 commented on docs/src/cahn_hilliard.md:
```diff
@@ -0,0 +1,99 @@
+  end
+end
+
+create_gif(soln, "CahnHillard_Rect.gif")
```
This is fixed now.

GeorgeR227 commented on examples/chemistry/brusselator_cuda.jl:
I've removed this file for now since this I don't want updating all the examples to hold up merging but I've made another branch based off this one to fixed the examples.

GeorgeR227 commented on examples/diff_adv/cht.jl:
```diff
@@ -0,0 +1,711 @@
+  end
+  return op
+end
+function simulate(mesh, operators, hodge = GeometricHodge())
```
Done, however we should clean up this example in another PR.

GeorgeR227 commented:
Yeah I'm looking at the history now and it seems like there have been some duplicated commits from even further back, I think from the last rebase. Take note of [3fd418e](https://github.com/AlgebraicJulia/Decapodes.jl/pull/225/commits/3fd418e9614e676de63f149f06bcd910901aeb5a) and [659ff27](https://github.com/AlgebraicJulia/Decapodes.jl/pull/225/commits/659ff27348240e07d81ce23c9fb3c9f2b7f4ee04). I'm thinking the history is messed up because it was already tangled with main's history. Notice that these new commits begin after the latest commit in main, namely [05376d9](https://github.com/AlgebraicJulia/Decapodes.jl/pull/225/commits/05376d9d24025fc947c1ad84e5ffe0e3c9735e4a).

GeorgeR227 commented:
@lukem12345 All tests are passing, including CUDA tests on HiPerGator. While the docs are failing I think this has been an ongoing issue and so should be addressed in the next PR.
